app/controllers/posts_controller.rb

```diff
@@ -103,15 +98,20 @@ def post_params
   #or admin within the specific post, user will be redirected to the post they just tried to edit.
   #Otherwise, the user and/or the admin has the rights to edit/delete said post
   def authorize_user
+    action = params['action']
     post = Post.find(params[:id])
 
-    if current_user.moderator?
-      return current_user.moderator!
-    else
-      unless current_user == post.user || current_user.admin?
-        flash[:alert] = "You must be an admin to do that"
-        redirect_to [post.topic, post]
-      end
+    # A post owner should be able to do everything.
+    return true if post.user == current_user
+
+    # Only admins may destroy a post.
+    if action == 'destroy' && !current_user.admin?
+      flash[:alert] = "You must be an admin to do that"
+      redirect_to [post.topic, post]
+    # Only admins and moderators may update a post.
+    elsif action == 'update' && !(current_user.moderator? || current_user.admin?)
+      flash[:alert] = "You must be an admin or a moderator to do that"
+      redirect_to [post.topic, post]
     end
   end
```

Comment:
The big picture here is that we want to react differently based on a duo:

* what role does the current user have?
* what is the current action?

I first "respond" to the latter by grabbing the action from the params: `action = params['action']`. In any request, the `params` hash will expose the controller's name and current action.

I then "respond" to the former question, in combination with the action's name:

* If the post we are dealing with belongs to the current user, just authorize everything. One may do this by having the filter method return `true`.
* If the action is "destroy", the specs say that only admins should be allowed to perform the deletion.
* If the action is "update", the specs say that only admins and moderators should be allowed to perform the update.

The overall logic is simpler, sticks to the specs, and the code easy to read.